File: app/src/main/java/com/spiderbiggen/randomchampionselector/ui/activities/LoaderActivity.java
package com.spiderbiggen.randomchampionselector.ui.activities;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.ddragon.DDragon;
import com.spiderbiggen.randomchampionselector.ddragon.ImageDescriptor;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.storage.database.DatabaseManager;
import com.spiderbiggen.randomchampionselector.storage.file.FileStorage;
import com.spiderbiggen.randomchampionselector.util.async.ProgressCallback;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.disposables.Disposable;

public class LoaderActivity extends AppCompatActivity implements ProgressCallback {

    private static final String TAG = LoaderActivity.class.getSimpleName();
    private DatabaseManager databaseManager;
    private DDragon dDragon;
    private List<Disposable> disposables = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_loader);

        ProgressBar progressBar = findViewById(R.id.progressBar);
        progressBar.setIndeterminate(true);

        databaseManager = DatabaseManager.getInstance();
        databaseManager.useContext(getApplicationContext());
        FileStorage.getInstance().setRootFromContext(this);
        dDragon = DDragon.getInstance();
        dDragon.setPreferences(PreferenceManager.getDefaultSharedPreferences(this));
        startLoading();
    }

    @Override
    protected void onDestroy() {
        for (Disposable disposable : disposables) {
            if (!disposable.isDisposed())
                disposable.dispose();
        }
        super.onDestroy();
    }

    private void startLoading() {
        disposables.add(dDragon.updateVersion(this::downloadChampions, this::catchError));
    }

    private void downloadChampions() {
        disposables.add(dDragon.getChampionList(this::handleChampionList, this::catchError));
    }

    private void handleChampionList(List<Champion> champions) {
        disposables.add(databaseManager.addChampions(champions));
        try {
            disposables.add(dDragon.verifyImages(champions, this, this::downloadMissingOrCorruptImages, this::catchError));
        } catch (IOException e) {
            catchError(e);
        }
    }

    private void catchError(Throwable t) {
        Log.e(TAG, "catchError: ", t);
        onError();
    }

    private void downloadMissingOrCorruptImages(List<ImageDescriptor> champions) {
        try {
            disposables.add(dDragon.downloadAllImages(champions, this, this::openMainScreen, this::catchError));
        } catch (IOException e) {
            catchError(e);
        }
    }

    private void openMainScreen() {
        Intent intent = new Intent(this, ListChampionsActivity.class);
        startActivity(intent);
    }

    @Override
    public void onProgressUpdate(Progress progress, int progressCount, int progressMax) {
        ProgressBar progressBar = findViewById(R.id.progressBar);

        if (progress == Progress.ERROR) {
            Drawable progressDrawable = progressBar.getIndeterminateDrawable().mutate();
            progressDrawable.setColorFilter(Color.RED, android.graphics.PorterDuff.Mode.SRC_IN);
            progressBar.setProgressDrawable(progressDrawable);
        }

        progressBar.setIndeterminate(progress.isIndeterminate());
        progressBar.setProgress(progressCount);
        progressBar.setMax(progressMax);

        TextView textView = findViewById(R.id.progressText);
        textView.setText(getString(progress.getStringResource(), progressCount, progressMax));
    }

    @Override
    public void finishExecution() {

    }

}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/views/activities/LoaderActivity.kt
package com.spiderbiggen.randomchampionselector.views.activities

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff.Mode
import android.net.ConnectivityManager
import android.os.Bundle
import androidx.annotation.UiThread
import com.spiderbiggen.randomchampionselector.data.PreferenceManager
import com.spiderbiggen.randomchampionselector.data.ddragon.DDragon
import com.spiderbiggen.randomchampionselector.model.IProgressCallback
import com.spiderbiggen.randomchampionselector.model.IProgressCallback.Progress
import kotlinx.android.synthetic.main.activity_loader.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import java.util.*


@ExperimentalCoroutinesApi
class LoaderActivity : AbstractActivity(), IProgressCallback {
    private var shouldRefresh: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!isNetworkAvailable()) {
            openListActivity()
            return
        }
        setContentView(com.spiderbiggen.randomchampionselector.R.layout.activity_loader)
        shouldRefresh = intent.getBooleanExtra(FORCE_REFRESH, false)
    }

    override fun onResume() {
        super.onResume()
        when {
            shouldRefresh || PreferenceManager.isOutdated -> {
                PreferenceManager.lastSync = -1
                updateData(this)
                onFinished()
            }
            else -> {
                verifyImages(this@LoaderActivity)
                onFinished()
            }
        }
    }

    private fun isNetworkAvailable(): Boolean {
        val connectivityManager =
            getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetworkInfo = connectivityManager.activeNetworkInfo
        return activeNetworkInfo?.isConnected == true
    }

    private fun onFinished() {
        PreferenceManager.lastSync = Date().time
        openListActivity()
    }

    private fun openListActivity() {
        val intent = Intent(this, ListChampionsActivity::class.java)
        startActivity(intent)
    }

    @UiThread
    override fun update(type: Progress, progress: Int, progressMax: Int) {
        if (type === Progress.ERROR) {
            val progressDrawable = progressBar.indeterminateDrawable.mutate()
            progressDrawable.setColorFilter(Color.RED, Mode.SRC_IN)
            progressBar.progressDrawable = progressDrawable
        }

        progressBar.isIndeterminate = type.indeterminate
        progressBar.progress = progress
        progressBar.max = progressMax

        val percent = if (progressMax == 0) 0f else (progress.toFloat() / progressMax) * 100
        progressText.text = getString(type.stringResource, percent)
    }

    private fun verifyImages(progress: IProgressCallback) {
        launch(Dispatchers.IO) {
            val champions = database.findChampionList()
            if (champions.isNullOrEmpty()) {
                updateData(progress)
            } else {
                val things = DDragon.verifyImages(champions, progress)
                DDragon.downloadAllImages(things, progress)
            }
        }
    }

    private fun updateData(progress: IProgressCallback) {
        launch(Dispatchers.IO) {
            progress.update(IProgressCallback.Progress.CHECKING_VERSION)
            val version = DDragon.getLastVersion()
            val champions = DDragon.getChampionList(version)
            database.addChampions(champions)
            val things = DDragon.verifyImages(champions, progress)
            if (!things.isEmpty()) {
                DDragon.downloadAllImages(things, progress)
            }
        }
    }

}
